ckanext/security/authenticator.py

```diff
@@ -13,54 +14,63 @@
 
 log = logging.getLogger(__name__)
 
+def get_request_ip_address(request):
+    """Retrieves the IP address from the request if possible"""
+    remote_addr = request.headers.get('X-Forwarded-For') or request.environ.get('REMOTE_ADDR')
+    if remote_addr == None:
+        log.critical('X-Forwarded-For header/REMOTE_ADDR missing from request.')
+
+    return remote_addr
+
+def get_login_throttle_key(request, user_name):
```

Comment:
I exposed this function so that the AJAX `login` method on the the controller could reuse the same logic as the `authenticate` method below. It centralises the lookup for the config key to decide on how what key should be used for brute force detection as well as the IP address check.